python/test/gluon/test_core.py

```diff
@@ -142,3 +145,145 @@ def test_warpgroup_mma(ASYNC):
+                     type_a: tl.constexpr, type_b: tl.constexpr):
+        DIV_FACTOR_A: tl.constexpr = 2 if type_a == "e2m1" else 1
+        DIV_FACTOR_B: tl.constexpr = 2 if type_b == "e2m1" else 1
+        PACKED_BLOCK_K_A: tl.constexpr = BLOCK_K // DIV_FACTOR_A
+        PACKED_BLOCK_K_B: tl.constexpr = BLOCK_K // DIV_FACTOR_B
+        SCALE_BLOCK_K: tl.constexpr = BLOCK_K // 32
+
+        a_unpacked_layout: ttgl.constexpr = ttgl.BlockedLayout([1, 16], [8, 8], [4, 1], [1, 0])
+        a_packed_layout: ttgl.constexpr = ttgl.BlockedLayout([1, 8], [8, 8], [4, 1], [1, 0])
+        a_layout: ttgl.constexpr = a_packed_layout if type_a == "e2m1" else a_unpacked_layout
+
+        a_scale_layout: ttgl.constexpr = ttgl.DistributedLinearLayout(
+            reg_bases=[], lane_bases=[[1, 0], [2, 0], [4, 0], [8, 0], [0, 1], [0, 2]], warp_bases=[[0, 0], [16, 0]],
+            block_bases=[], shape=[32, 4])
+
+        b_unpacked_layout: ttgl.constexpr = ttgl.BlockedLayout([1, 16], [32, 2], [4, 1], [1, 0])
+        b_packed_layout: ttgl.constexpr = ttgl.BlockedLayout([1, 8], [16, 4], [4, 1], [1, 0])
+        b_layout: ttgl.constexpr = b_packed_layout if type_b == "e2m1" else b_unpacked_layout
+
+        b_scale_layout: ttgl.constexpr = ttgl.DistributedLinearLayout(
+            reg_bases=[], lane_bases=[[1, 0], [2, 0], [4, 0], [8, 0], [0, 1], [0, 2]], warp_bases=[[16, 0], [0, 0]],
+            block_bases=[], shape=[32, 4])
+
+        mfma_layout: ttgl.constexpr = ttgl.amd.AMDMFMALayout(version=4, warps_per_cta=[2, 2], tiles_per_warp=[1, 1],
+                                                             instr_shape=[16, 16], transposed=True)
+
+        zero = ttgl.zeros([BLOCK_M, BLOCK_N], dtype=ttgl.float32, layout=mfma_layout)
+
+        a_offsets = ttgl.arange(0, BLOCK_M, layout=ttgl.SliceLayout(1, a_layout))[:, None] * stride_am + \
+                    ttgl.arange(0, PACKED_BLOCK_K_A, layout=ttgl.SliceLayout(0, a_layout))[None, :] * stride_ak
+        a = ttgl.amd.cdna4.buffer_load(a_base, a_offsets)
+        a = ttgl.convert_layout(a, ttgl.DotOperandLayout(operand_index=0, parent=mfma_layout, k_width=16))
+
+        b_offsets = ttgl.arange(0, PACKED_BLOCK_K_B, layout=ttgl.SliceLayout(1, b_layout))[:, None] * stride_bk + \
+                    ttgl.arange(0, BLOCK_N, layout=ttgl.SliceLayout(0, b_layout))[None, :] * stride_bn
+        b = ttgl.amd.cdna4.buffer_load(b_base, b_offsets)
+        b = ttgl.convert_layout(b, ttgl.DotOperandLayout(operand_index=1, parent=mfma_layout, k_width=16))
+
+        if a_scale is not None:
+            a_scale_offsets = ttgl.arange(0, BLOCK_M, layout=ttgl.SliceLayout(1, a_scale_layout))[:, None] * SCALE_BLOCK_K + \
+                              ttgl.arange(0, SCALE_BLOCK_K, layout=ttgl.SliceLayout(0, a_scale_layout))[None, :]
+            a_scale = ttgl.amd.cdna4.buffer_load(a_scale, a_scale_offsets)
+        else:
+            a_scale = ttgl.full([BLOCK_M, SCALE_BLOCK_K], 127, dtype=ttgl.int8, layout=a_scale_layout)
```

Comment:
We might want to support broadcasting to make such constant scale cases simpler to write?

Reply:
I don't quite follow: what do you mean by "broadcasting" here? I thought `ttgl.full` will just give a constant tensor with this shape and layout which is exactly what we need here.

Reply:
I meant that we can just provide a scalar value instead of mandating a full tensor here. Internally we can broadcast. You can see for example [`broadcast_impl_value`](https://github.com/triton-lang/triton/blob/69d58b5c68fa942c668f94bf5a2e1f137b205ed8/python/triton/experimental/gluon/language/_semantic.py#L113) and [`buffer_load`](https://github.com/triton-lang/triton/blob/69d58b5c68fa942c668f94bf5a2e1f137b205ed8/python/triton/experimental/gluon/language/amd/cdna3/__init__.py#L74). Though those are more elementwise broadcast; here we might need to figure out the scale layout internally.

Reply:
I think broadcasting from a scalar tensor requires a tensor with the desired shape. however we don't have a tensor with the scale shape here.

This actually mimics handling of none scale in the `AccelerateAMDMatmul` pass:

https://github.com/triton-lang/triton/blob/69d58b5c68fa942c668f94bf5a2e1f137b205ed8/third_party/amd/lib/TritonAMDGPUTransforms/AccelerateAMDMatmul.cpp#L1015-L1017

Reply:
the tricky part is we need to construct a tensor with the matched layout even for a constant value, either using full or constant+broadcast as dot_scaled requires a tensor scale. we still need to determine things like of size per threads, essentially following the logic of `AccelerateAMDMatmul` pass. it is possible, but this will complicate the internal logic of the `mfma_scaled`.

Reply:
Okay. I think we need to have some utilities in Gluon to make such job easier to deduce linear layout from other layouts etc. So fine to leave this as a future improvement then.